### PR TITLE
Add support for cross compiling to ESP MCUs

### PR DESCRIPTION
**Pull request by plietar**
###### Motivation for this change

Provide a toolchain for Espressif's Xtensa-based MCUs. This uses Espressif's
GCC/binutils/newlib forks, as well as applies a variant specific overlay.

The toolchain can be used both with the ESP-IDF sdk or "bare-metal" (eg. using https://github.com/taktoa/esp32-baremetal).

This is a rebased / slightly improved version of #96131.

While the PR should support the esp32, esp32s2 and esp8266, I only have the first one to try. I should be able to get my hands on an esp8266 this week to have a try at it. It could also support the newly announced esp32s3, but since there are no publicly available boards yet I've omitted it for `lib/systems` for now.

There are still a couple of issues / potential changes to be made.
- I used a `isXtensa` predicate in a bunch of places. This isn't very accurate, since there could be some non-esp xtensa chips. A `isEspressif` predicate wouldn't work either since espressif has some RISC-V cores too. `isXtensa && isEspressif` is a bit verbose. I'm inclined to keep using `isXtensa` until another xtensa chip comes along and/or support for a RISC-V ESP core is added (for instance, I think the newlib package would be shared).
- The way the overlay package is downloaded and distributed is probably bad. I don't know what the best practice is for this.
- I've copied the gcc flags and newlib configuration flags from [Espressif's crosstool-ng config](https://github.com/espressif/crosstool-NG/blob/esp-1.23.x/samples/xtensa-esp32-elf/crosstool.config).
  - `-mlongcall` is required otherwise I get linker errors on anything non-trivial.
  - `--enable-newlib-long-time_t` (ie. 32bit time_t) is required by the default ESP-IDF configuration, but there's a flag in the IDF to use if the toolchain uses 64bit time_t.
  - The rest, I think, are strictly a matter of preference and space/time tradeoff.
- (Off-topic) `or1k` and `vc4` have their own newlib package definitions that could be merged with the default `development/misc/newlib` one. I'd originally done this, but decided to leave it for another time to keep this PR on-topic.
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by andir** [comment]
@plietar do you mind rebasing this? Based on the feedback in the original PR this should still be very welcome. We can try to ping those with domain knowledge again to see what they say: @Ericson2314 @Mic92

File: lib/systems/examples.nix [hunk 1]
```diff
@@ -137,6 +137,19 @@ rec {
...
+  esp8266 = {
+    config = "xtensa-lx106-elf";
+    libc = "newlib";
+  };
+  esp32 = {
+    config = "xtensa-esp32-elf";
+    libc = "newlib";
+  };
+  esp32s2 = {
+    config = "xtensa-esp32s2-elf";
```
Comment (Ericson2314) [inline lib/systems/examples.nix:149]: this looks like a very non-standard triple to me? Why not `esp32s2-unknown-none-elf`?

File: pkgs/development/tools/misc/binutils/default.nix [hunk 2]
```diff
@@ -125,7 +135,10 @@ stdenv.mkDerivation {
     # RUNPATH instead of RPATH on binaries.  This is important because
     # RUNPATH can be overriden using LD_LIBRARY_PATH at runtime.
     "--enable-new-dtags"
-  ] ++ lib.optionals gold [ "--enable-gold" "--enable-plugins" ];
+  ] ++ (lib.optionals gold [ "--enable-gold" "--enable-plugins" ])
+
+  # Git-based builds default to building both binutils and gdb at once. 
+  ++ (lib.optional stdenv.targetPlatform.isXtensa "--disable-gdb");
```
Comment (Mic92) [inline pkgs/development/tools/misc/binutils/default.nix:141]: We can pass this option by default, no?

**Comment by siraben** [comment]
@plietar please resolve the merge conflicts. I was actually looking into adding xtensa cross-compilation support today, and glad you already got it started!

**Comment by alufers** [comment]
Hi! What's the status of this pull request? I can rebase this PR and create a new one, if you don't have the time.

**Comment by veprbl** [comment]
Rebased and addressed comments in #144452

**Comment by lheckemann** [comment]
Superseded by #144452

**Pull request closed without merging**